### PR TITLE
adding compression for OIDC cookies

### PR DESCRIPTION
**Pull request by universam1**
This is the second part follow up of #1089 to properly deal with large JWT token and cookies.

Compression is trading computing with bandwidth, however we consider a 🐛 in conjunction with _AWS Application LoadBalancer_ which has a hard limit on header/cookies of 16kb, that we exceed right now! Skipper could with this PR not only solve this issue but improve the roundtrip through significantly less payload!

Confirmed it solves that describedcurrent problem, tested in integration setup:
`[APP]time="2020-01-29T13:39:51Z" level=debug msg="cookie compressed: 27695 to 11147"`

Also, since the payload for **requests (cookies) are reduced by a factor of ~2.5** the overall performance is measurable improved!

Tests showed that compression on the cookies pays off from >200 bytes payload which is granted for OIDC type payload:
```
=== RUN   TestRatio
raw: 100 comp: 124
raw: 200 comp: 185
raw: 300 comp: 260
raw: 13900 comp: 5016
```

Leveraging here `sync.Pools` that provide safe parallelization, performance >2x and through salvaging the writer also drops memory footprint by factor 8 compared to new instances of `flate.Writer` - benchmarks included, see this branch for all comparisons: https://github.com/universam1/skipper/blob/oidcCompressAll/filters/auth/oidc_test.go#L839

```
running tool: /usr/local/bin/go test -benchmem -run=^$ github.com/zalando/skipper/filters/auth -bench ^(Benchmark_deflatePoolCompressor)$ -v

goos: darwin
goarch: amd64
pkg: github.com/zalando/skipper/filters/auth
...
Benchmark_deflatePoolCompressor/ST/comp/flate/max-16         	    7384	    158146 ns/op	         2.52 /1-ratio	       389 byte	   55746 B/op	      14 allocs/op
Benchmark_deflatePoolCompressor/MT/comp/flate/max-16         	   43186	     29916 ns/op	   49864 B/op	      13 allocs/op
...
Benchmark_deflatePoolCompressor/ST/decomp/flate/max-16       	    7098	    157146 ns/op	         2.52 /1-ratio	       389 byte	   55122 B/op	      14 allocs/op
Benchmark_deflatePoolCompressor/MT/decomp/flate/max-16       	   40233	     29047 ns/op	   49809 B/op	      13 allocs/op
```
We might conclude that 40k authentications == fresh cookies/second are reasonable on a server.

`Deflate` compression is chosen after tests with similar algorithms, providing the best balance of simple implementation, speed and ratio.

```
BenchmarkCompressSerial/comp/flate-pool/default-16         	    1230	    918036 ns/op	      5229 byte	         3.40 ratio	  136876 B/op	      26 allocs/op
BenchmarkCompressSerial/comp/flate-pool/max-16             	    1232	   1021302 ns/op	      5151 byte	         3.45 ratio	  144771 B/op	      26 allocs/op
BenchmarkCompressSerial/comp/flate-pool/min-16             	    1934	    622571 ns/op	      5866 byte	         3.03 ratio	  153949 B/op	      24 allocs/op
BenchmarkCompressSerial/comp/flate/default-16              	    1053	   1205067 ns/op	      5229 byte	         3.40 ratio	  933267 B/op	      42 allocs/op
BenchmarkCompressSerial/comp/flate/max-16                  	     936	   1297204 ns/op	      5151 byte	         3.45 ratio	  933266 B/op	      42 allocs/op
BenchmarkCompressSerial/comp/flate/min-16                  	    1203	   1079002 ns/op	      5866 byte	         3.03 ratio	 1326434 B/op	      42 allocs/op
BenchmarkCompressSerial/comp/gzip-pool/default-16          	    1246	    957243 ns/op	      5247 byte	         3.39 ratio	  150496 B/op	      29 allocs/op
BenchmarkCompressSerial/comp/gzip-pool/max-16              	    1140	   1038045 ns/op	      5169 byte	         3.44 ratio	  153704 B/op	      29 allocs/op
BenchmarkCompressSerial/comp/gzip-pool/min-16              	    1824	    643649 ns/op	      5884 byte	         3.02 ratio	  171982 B/op	      27 allocs/op
BenchmarkCompressSerial/comp/gzip/default-16               	     996	   1258406 ns/op	      5247 byte	         3.39 ratio	  937425 B/op	      45 allocs/op
BenchmarkCompressSerial/comp/gzip/max-16                   	     909	   1347210 ns/op	      5169 byte	         3.44 ratio	  937425 B/op	      45 allocs/op
BenchmarkCompressSerial/comp/gzip/min-16                   	    1082	   1070842 ns/op	      5884 byte	         3.02 ratio	 1330593 B/op	      45 allocs/op
BenchmarkCompressSerial/comp/zlib/default-16               	     943	   1257686 ns/op	      5235 byte	         3.40 ratio	  936752 B/op	      48 allocs/op
BenchmarkCompressSerial/comp/zlib/max-16                   	     889	   1358716 ns/op	      5157 byte	         3.45 ratio	  936752 B/op	      48 allocs/op
BenchmarkCompressSerial/comp/zlib/min-16                   	    1095	   1110003 ns/op	      5872 byte	         3.03 ratio	 1329919 B/op	      48 allocs/op
BenchmarkCompressSerial/decomp/flate-pool/default-16       	    1278	    931506 ns/op	      5229 byte	         3.40 ratio	  145145 B/op	      26 allocs/op
BenchmarkCompressSerial/decomp/flate-pool/max-16           	    1184	   1024162 ns/op	      5151 byte	         3.45 ratio	  147167 B/op	      26 allocs/op
BenchmarkCompressSerial/decomp/flate-pool/min-16           	    1903	    609047 ns/op	      5866 byte	         3.03 ratio	  149431 B/op	      24 allocs/op
BenchmarkCompressSerial/decomp/flate/default-16            	     979	   1209546 ns/op	      5229 byte	         3.40 ratio	  933265 B/op	      42 allocs/op
BenchmarkCompressSerial/decomp/flate/max-16                	     902	   1306409 ns/op	      5151 byte	         3.45 ratio	  933265 B/op	      42 allocs/op
BenchmarkCompressSerial/decomp/flate/min-16                	    1158	   1054071 ns/op	      5866 byte	         3.03 ratio	 1326433 B/op	      42 allocs/op
BenchmarkCompressSerial/decomp/gzip-pool/default-16        	    1252	    973155 ns/op	      5247 byte	         3.39 ratio	  151017 B/op	      29 allocs/op
BenchmarkCompressSerial/decomp/gzip-pool/max-16            	    1156	   1039359 ns/op	      5169 byte	         3.44 ratio	  151879 B/op	      29 allocs/op
BenchmarkCompressSerial/decomp/gzip-pool/min-16            	    1914	    645426 ns/op	      5884 byte	         3.02 ratio	  180432 B/op	      28 allocs/op
BenchmarkCompressSerial/decomp/gzip/default-16             	    1065	   1231097 ns/op	      5247 byte	         3.39 ratio	  937426 B/op	      45 allocs/op
BenchmarkCompressSerial/decomp/gzip/max-16                 	     910	   1322129 ns/op	      5169 byte	         3.44 ratio	  937425 B/op	      45 allocs/op
BenchmarkCompressSerial/decomp/gzip/min-16                 	    1088	   1082934 ns/op	      5884 byte	         3.02 ratio	 1330594 B/op	      45 allocs/op
BenchmarkCompressSerial/decomp/zlib/default-16             	     949	   1302399 ns/op	      5235 byte	         3.40 ratio	  936751 B/op	      48 allocs/op
BenchmarkCompressSerial/decomp/zlib/max-16                 	     876	   1336452 ns/op	      5157 byte	         3.45 ratio	  936752 B/op	      48 allocs/op
BenchmarkCompressSerial/decomp/zlib/min-16                 	    1123	   1109236 ns/op	      5872 byte	         3.03 ratio	 1329919 B/op	      48 allocs/op
```

### CODE REVIEW

**Comment by szuecs** [comment]
Thanks for the PR, I will review it next week

File: filters/auth/oidc_test.go [hunk 1]
```diff
@@ -63,65 +64,66 @@ func (jar *insecureCookieJar) Cookies(u *url.URL) []*http.Cookie {
 	return jar.store[u.Hostname()]
 }
 
+var s = `{
```
Comment (szuecs) [inline filters/auth/oidc_test.go]: `testOpenIDConfig` instead of `s`

File: filters/auth/oidc_test.go [hunk 2]
```diff
@@ -826,3 +828,49 @@ func TestChunkAndMergerCookie(t *testing.T) {
 		})
 	}
 }
+
+var cookieCompressRuns = []struct {
+	name string
+	fun  cookieCompression
```
Comment (szuecs) [inline filters/auth/oidc_test.go]: `fun` -> `compressor`

**Comment by szuecs** [comment]
besides the renaming lgtm

File: filters/auth/oidc_test.go (hunk 1 above)
Comment (universam1) [reply filters/auth/oidc_test.go]: done

File: filters/auth/oidc_test.go (hunk 2 above)
Comment (universam1) [reply filters/auth/oidc_test.go]: changed

**Comment by universam1** [comment]
@szuecs remarks applied, ready for your merge

**Comment by szuecs** [comment]
:+1:

**Comment by aryszka** [comment]
thank you very much!

**Comment by aryszka** [comment]
:+1:

**Comment by universam1** [comment]
@szuecs @aryszka Kindly request for a release if possible? Thank you!

**Comment by szuecs** [comment]
@universam1 the release is automated only for Zalando employees, because we are only allowed to run builds from us in the build environment. We jsut merged another PR that got released, so v0.11.35 has your changes. I will update the release to show your change in it, too.
Sorry for the inconvenience, but helps us to work efficiently.

**Pull request merged**